AAXtoMP3

```diff
@@ -1,166 +1,305 @@
 #!/usr/bin/env bash
 
-set -o errexit -o noclobber -o nounset -o pipefail
 
-codec=libmp3lame
-extension=mp3
-mode=chaptered
-authcode=".authcode";
-if [ -z ${HOME+x} ] && ! [ -z ${USERPROFILE+x} ]; then HOME="$USERPROFILE"; fi
-authcodeDirs="${HOME}/ ./"
-GREP=$(grep --version | grep -q GNU && echo "grep" || echo "ggrep")
+# ========================================================================
+# Command Line Options
 
-if ! [[ $(type -P "$GREP") ]]; then
-    echo "$GREP (GNU grep) is not in your PATH"
-    echo "Without it, this script will break."
-    echo "On macOS, you may want to try: brew install grep"
-    exit 1
-fi
+# Usage Synopsis.
+usage=$'\nUsage: AAXtoMP3.sh [--flac] [--aac] [--opus ] [--single] [-e:m4a] [-e:m4b]\n[--authcode <AUTHCODE>] [--output_dir <PATH>] {FILES}\n'
+codec=libmp3lame            # Default encoder.
+extension=mp3               # Default encoder extention.
+mode=chaptered              # Multi file output
+auth_code=                  # Required to be set via file or option.
+targetdir=                  # Optional output location.  Note default is basedir of AAX file.
+DEBUG=0                     # Default off, If set extremely verbose output.
+container=                  # Just in case we need to change the container.  Used for M4A to M4B
 
+# -----
+# Code tip Do not have any script above this point that calls a function or a binary.  If you do
+# the $1 will no longer be a ARGV element.  So you should only do basic variable setting above here.
+#
+# Process the command line options.  This allows for un-ordered options. Sorta like a getops style
+while true; do
+  case "$1" in 
+                      # Flac encoding
+    -f | --flac       ) codec=flac;    extension=flac; mode=single ;                shift ;;        
+                      # Apple m4a music format.
+    -a | --aac        ) codec=copy;    extension=m4a; mode=single ;                 shift ;;        
+                      # Ogg Format
+    -o | --opus       ) codec=libopus; extension=ogg;                               shift ;;        
+                      # If appropriate use only a single file output.
+    -s | --single     ) mode=single;                                                shift ;;        
+                      # This is the same as --single option.
+    -e:mp3            ) codec=libmp3lame; extension=mp3; mode=single;               shift ;;
+                      # Identical to --acc option.
+    -e:m4a            ) codec=copy; extension=m4a; mode=single;                     shift ;;
+                      # Similiar to --aac but specific to audio books.
+    -e:m4b            ) codec=copy; extension=m4a; mode=single; container=m4b       shift ;;
```

Comment:
Can you unset `mode=single` in this param? I think m4b is better with split files.

Reply:
How about this.  I add a --chaptered switch to force split files.

The original design by Apple was that a single m4b file contains all of the chapters in one file. Since the m4b already internally has chapter markers.

Which I added in the next code version.  There will be conditions on this documented in the readme.  As this flag will break --flac selection.  Note flac splitting is a known bug in ffmpeg.

Reply:
Sure, it'd be a better solution, thanks.